### PR TITLE
SearChain package reproduced using Llama_index library ​

### PR DESCRIPTION
**Pull request by DJC-GO-SOLO**
# Description

This is an incremental contribution. We reproduce the [SearChain](https://github.com/xsc1234/Search-in-the-Chain) framework using Llama_index. This helps more people use it.

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

We provide an example notebook. And we have tested our code and the relevant data can be found in the  [SearChain](https://github.com/xsc1234/Search-in-the-Chain) code base.
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-packs/llama-index-packs-searchain/llama_index/packs/searchain/base.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+                if ':' in query_item:
+                    query_item = query_item[1:]
+                #print('solving: '+query_item)
+                if not _have_seen_or_not(self.crossencoder,query_item,query_seen_list,query_type):
+                    now_reference = {}
+                    query_seen_list.append(query_item)
+                    response = str(self.query_engine.query(query_item))
+#                   print('query_item:'+query_item)
+#                   print('top1_passage:'+response)
+#                   print(response)
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-searchain/llama_index/packs/searchain/base.py:114]: we can remove these debug prints

File: llama-index-packs/llama-index-packs-searchain/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+# Feel free to un-skip examples, and experimental, you will just need to
+# work through many typos (--write-changes and --interactive will help)
+skip = "*.csv,*.html,*.json,*.jsonl,*.pdf,*.txt,*.ipynb"
+
+# [tool.llamahub]
+# contains_example = false
+# import_path = "<import-path>"
+
+# [tool.llamahub.class_authors]
+# CLASS = "github-username"
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-searchain/pyproject.toml:17]: you can fill in these two llamahub sections

File: llama-index-packs/llama-index-packs-searchain/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+version = "0.1.0"
+description = "llama-index packs searchain integration"
+authors = ["Your Name <you@example.com>"]
+license = "MIT"
+readme = "README.md"
+packages = [{include = "llama_index/"}]
+
+[tool.poetry.dependencies]
+python = ">=3.8.1,<4.0"
+llama-index-core = "^0.10.0"
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-searchain/pyproject.toml:37]: lets be sure to add any external dependencies here (I see at least transformers, torch, and sentence transformers all being used)

File: llama-index-packs/llama-index-packs-searchain/README.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# LlamaIndex Packs Integration: Searchain
+This LlamaPack implements short form the [SearChain paper by Xu et al..](https://arxiv.org/abs/2304.14732)
+
+You can see its use case in the examples folder.
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-searchain/README.md:4]: We can paste the example from the notebook here tbh (and we should also include an example of installing and/or downloading the pack)

File: llama-index-packs/llama-index-packs-searchain/examples/searchain.ipynb
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-searchain/examples/searchain.ipynb:1]: Should we provide explicit links on how to get/download the models needed?

Comment (DJC-GO-SOLO) [reply llama-index-packs/llama-index-packs-searchain/examples/searchain.ipynb:1]: Yes, they are included in the readme.md file.

**Review by logan-markewich — APPROVED** [review]

**Comment by DJC-GO-SOLO** [comment]
Could you please provide me with some guidance on how to successfully navigate through this test?

**Comment by nerdai** [comment]
> Could you please provide me with some guidance on how to successfully navigate through this test?

Hey you need a couple more BUILD files. You can create them manually, or you could add me to your fork and I can make the push from my end.

`llama_index/packs/searchain/BUILD`
```
python_sources()
```

`tests/BUILD`
```
python_tests()
```

Once those are added, that should pass checks.

**Comment by DJC-GO-SOLO** [comment]
> > Could you please provide me with some guidance on how to successfully navigate through this test?
> 
> Hey you need a couple more BUILD files. You can create them manually, or you could add me to your fork and I can make the push from my end.
> 
> `llama_index/packs/searchain/BUILD`
> 
> ```
> python_sources()
> ```
> 
> `tests/BUILD`
> 
> ```
> python_tests()
> ```
> 
> Once those are added, that should pass checks.

Thank you very much, I'll try it by myself first.

**Comment by nerdai** [comment]
Thanks @DJC-GO-SOLO -- merging now!

**Pull request merged**